src/server.rs

```diff
@@ -303,51 +393,149 @@ impl ServerState {
+                    }
+                }
             }
+
         }
-        /*
-        match action {
-            RequestAction::Connect => {
-                self.players.iter().for_each(|player| {
-                    assert_eq!(true, player.addr != addr && player.player_name != player_name);
-                });
+    }
 
-                self.players.push(Player::new(player_name, addr));
-            },
-            RequestAction::Ack                 => {},
-            RequestAction::Disconnect          => {},
-            RequestAction::JoinGame            => {},
-            RequestAction::ListPlayers         => {},
-            RequestAction::ChatMessage(String) => {},
-            RequestAction::None                => {},
+    // XXX
+    // Right now we'll be constructing all client Update packets for _every_ game slot.
+    fn construct_client_updates(&mut self) -> Result<Option<Vec<(SocketAddr, Packet)>>, Box<Error>> {
+        let mut client_updates: Vec<(SocketAddr, Packet)> = vec![];
+
+        if self.game_slots.len() == 0 {
+            return Ok(None);
+        }
+
+        // For each game slot, determine if each player has unread messages based on last_acked_msg_seq_num
+        // TODO: POOR PERFORMANCE BOUNTY
+        for slot in self.game_slots.values() {
+            for player_id in &slot.player_ids {
+                let mut unsent_messages: Vec<BroadcastChatMessage> = vec![];
+                let opt_player = self.players.get(&player_id);
+                match opt_player {
+                    Some(player_info) => {
+
+                        // MESSAGE UPDATING
+                        match player_info.last_acked_msg_seq_num {
+                            Some(last_acked_msg_seq_num) => {
+                                if !slot.messages.is_empty() {
+                                    let mut message_iter = slot.messages.iter();
+                                    let mut message = message_iter.next().unwrap(); // From front to back
+                                    let mut amount_to_consume = 0; // Skip over these messages since we've already acked them
+
+                                    if last_acked_msg_seq_num - message.seq_num > 0{
+                                        amount_to_consume = (last_acked_msg_seq_num - message.seq_num) % (MAX_AGE_CHAT_MESSAGES as u64);
+                                    }
+
+                                    let mut message_iter = message_iter.skip(amount_to_consume as usize);
+
+                                    while let Some(message) = message_iter.next() {
+                                        if message.player_id != player_info.player_id {
+                                            unsent_messages.push(BroadcastChatMessage{
+                                                chat_seq:    Some(message.seq_num),
+                                                player_name: message.player_name.clone(), // XXX is clone really the best way to do this?
```

Comment:
Aside from sending a Player ID and having the client look that up, I don't think we can do better than cloning the player name.

Reply:
👍